src/compiler/rust.rs

```diff
@@ -1184,6 +1427,239 @@ impl pkg::ToolchainPackager for RustToolchainPackager {
+        use tempfile;
+
+        let temp_dir = tempfile::tempdir()
+            .chain_err(|| "Could not create temporary file for rlib output")?;
+        let temp_rlib = temp_dir.path().join("x.rlib");
+
+        let mut cmd = process::Command::new(&executable);
+        cmd
+            .arg("--crate-type=rlib")
+            .arg("-o").arg(&temp_rlib)
+            .arg("-")
+            .env_clear()
+            .envs(ref_env(env_vars));
+
+        let process::Output { status, stdout, stderr } = cmd.output()?;
+
+        if !status.success() {
+            bail!("Failed to compile a minimal rlib with {}", executable.display())
+        }
+        if !stdout.is_empty() {
+            bail!("rustc stdout non-empty when compiling a minimal rlib: {:?}", String::from_utf8_lossy(&stdout))
+        }
+        if !stderr.is_empty() {
+            bail!("rustc stderr non-empty when compiling a minimal rlib: {:?}", String::from_utf8_lossy(&stderr))
+        }
+
+        // The goal of this cache is to avoid repeated lookups when building a single project. Let's budget 3MB.
+        // Allowing for a 100 byte path, 50 dependecies per rlib and 20 characters per crate name, this roughly
+        // approximates to `path_size + path + vec_size + num_deps * (systemtime_size + string_size + crate_name_len)`
+        //                 `   3*8    +  100 +   3*8    +    50    * (      8         +     3*8     +       20      )`
+        //                 `2748` bytes per crate
+        // Allowing for possible overhead of up to double (for unused space in allocated memory), this means we
+        // can cache information from about 570 rlibs - easily enough for a single project.
+        const CACHE_SIZE: u64 = 3*1024*1024;
+        let cache = lru_cache::LruCache::with_meter(CACHE_SIZE, DepsSize);
+
+        let rlib_dep_reader = RlibDepReader { cache: Mutex::new(cache), executable };
+        if let Err(e) = rlib_dep_reader.discover_rlib_deps(env_vars, &temp_rlib) {
+            bail!("Failed to read deps from minimal rlib: {}", e)
+        }
+
+        Ok(rlib_dep_reader)
+    }
+
+    fn discover_rlib_deps(&self, env_vars: &[(OsString, OsString)], rlib: &Path) -> Result<Vec<String>> {
+        let rlib_mtime = fs::metadata(&rlib).and_then(|m| m.modified()).chain_err(|| "Unable to get rlib modified time")?;
+
+        {
+            let mut cache = self.cache.lock().unwrap();
+            if let Some(deps_detail) = cache.get(rlib) {
+                if rlib_mtime == deps_detail.mtime {
+                    return Ok(deps_detail.deps.clone())
+                }
+            }
+        }
+
+        trace!("Discovering dependencies of {}", rlib.display());
+
+        let mut cmd = process::Command::new(&self.executable);
+        cmd.args(&["-Z", "ls"]).arg(&rlib)
+            .env_clear()
+            .envs(ref_env(env_vars))
+            .env("RUSTC_BOOTSTRAP", "1"); // TODO: this is fairly naughty
```

Comment:
Is this something we expect will ever become stable? If not it's probably a necessary evil, but it's a pretty minor thing.